csrc/kernel_cache.cpp

```diff
@@ -755,6 +754,44 @@ std::unordered_map<Val*, const ArgAbstract*> FusionKernelRuntime::
       " inputs but expected ",
       segmented_fusion_->inputs().size());
 
+  // map val to segment_id where arg is lastly used
+  std::unordered_map<Val*, int> last_used_segment_map;
+  // map segment_id to vector of fusion vals lastly used at this segment
+  std::unordered_map<int, std::vector<Val*>> vals_last_used_at_segment;
+  const int n_groups = runtime_workspace_.group_run_order.size();
+  // only need to set lifetime of vals if there are more than 3 groups
+  if (n_groups >= 3) {
+    // start from the 2nd group, since the input of the first group is always
+    // the global input and its outputs are always used by at least one of the
+    // following groups
+    for (int group_id = 1; group_id < n_groups; ++group_id) {
+      auto group_to_run = runtime_workspace_.group_run_order.at(group_id);
+      // set/update life of vals in inputs of this group
+      for (auto val : group_to_run->inputs()) {
+        // skip fusion inputs
```

Comment:
Please add this comment:

```
// At this point, the fusion is still the complete, unsegmented fusion, and if a Val is a fusion input, it means it's a user input.

Reply:
I understand `if a Val is a fusion input, it means it's a user input.` but why `the fusion is still the complete, unsegmented fusion`? The fusion should already be segmented, right? or whether is is segmented or not it doesn't influence the test of `val->isFusionInput()` becase `input->setIsFusionInput(true);` was called from `void Fusion::addInput(Val* input)`

Reply:
Thanks for the clarification. Here the code is looping over `SegmetnedFusion` using `auto group_to_run = group_run_order_.at(group_id);` and check its input `for (auto val : group_to_run->inputs())`. then skip the input `val` which is also a fusion input `val->isFusionInput()`.

In your comments, `a val that is originally just an intermediate val may become a fusion input `, here you are referecing to `segmented fusion input`, right? It can never becomes a `isFusionInput()` which only returns true for user provided input.

Reply:
You're right. I thought we were not making a copy of the original fusion for each segment, but we do:

https://github.com/NVIDIA/Fuser/blob/main/csrc/fusion_segmenter.cpp#L1437-L1438

So, we modify fusion inputs and outputs in the copy, so the original fusion is never changed. 

Creating a complete deep copy for each segment seems wasteful. A tensor may appear in multiple segments when used by multiple segments, so once scheduled in a segment, it would appear already scheduled in another segment, which conflicts with an assumption that each segment starts with tensors with no prior scheduling. We could probably cancel the scheduling of such tensors, or more fundamentally we don't need to schedule fusion inputs at all, so these constraints are more like artificial rather than fundamental.